### DISCUSSION TITLE
How can I find_package from dependent crate?

### DISCUSSION THREAD

**Discussion by lemoncmd** [alexcrichton/cmake-rs#222] [Q&A]
I have two crates with cmake, and one crate A depends on another crate B.

How can a cmake project in crate A find_package from a cmake project in crate B?

```
crate A -> cmake project A
| [dependencies]    | find_package?
v                   v
crate B -> cmake project B
```

**Comment by sksat** [comment]
The actual crate source directory is managed by Cargo, so we can't know where it is. Unless you explicitly specify a `path`, it will reside deep in a directory such as `~/.cargo/regostry/` or `~/.cargo/git`. And there is no official way to know the actual location of dependencies from build script.

There are two workarounds for this problem.

1. Propagate `CARGO_MANIFEST_DIR` using `links` manifest key
In the build script, the actual location of self crate source directory can be obtained using `CARGO_MANIFEST_DIR`.
We can use [The links Manifest Key](https://doc.rust-lang.org/cargo/reference/build-scripts.html#the-links-manifest-key) to propagate the directory to the upper crate.
However, `links` is designed for -sys crate etc. and has some limitations.

2. Use cargo-metadata
[cargo-metadata(1)](https://doc.rust-lang.org/cargo/commands/cargo-metadata.html) can be used to retrieve various metadata, including the actual location of the source directory.
ex: `dirname $(cargo metadata --format-version 1 | jq '.packages[] | select(.name == "B")' | .manifest_path)`
For use from within the build script, the [cargo_metadata](https://crates.io/crates/cargo_metadata) crate will be useful.

**Comment by tgross35** [comment]
Thanks @sksat for the great answer. I don't think there is anything specific to cmake-rs here, so I'll move this to a discussion.
